### ISSUE TITLE
Searching by custom product attribute doesn't work because of arbitraty backend model check

### ISSUE THREAD

**Issue by bartek9007** [Smile-SA/elasticsuite#943]
When indexing or performing search, there is a check if backend model (if exists) of attribute is in array of hardcoded backend models and if it's not it is not indexed and we are unable to use it in search.

### Preconditions
Magento Version : CE 2.2.4

ElasticSuite Version : 2.5.1

Environment : Developer

Third party modules : No

### Steps to reproduce
1. Add programatically attribute to be searchable and has custom backend model, according to your documentation
2. Set its value for product and save product.
3. Frontend - search by its value, but no luck.

### Expected result
1. It should be found, and if it's only search result I should have been redirected to its product page.

### Actual result
1. Nothing found.

Actually, I have found place where something strange is happening. In `vendor/smile/elasticsuite/src/module-elasticsuite-catalog/Model/Eav/Indexer/Fulltext/Datasource/AbstractAttributeData.php` in method 
`private function canIndexAttribute(AttributeInterface $attribute)
    {
        // 'price' attribute is declared as nested field into the indices file.
        $canIndex = $attribute->getBackendType() != 'static' && $attribute->getAttributeCode() !== 'price';

        if ($canIndex && $attribute->getBackendModel()) {
            $canIndex = in_array($attribute->getBackendModel(), $this->indexedBackendModels);
        }

        return $canIndex;
    }`

there is strange check performed. Backend model must be one of 
`protected $indexedBackendModels = [
        'Magento\Eav\Model\Entity\Attribute\Backend\ArrayBackend',
        'Magento\Eav\Model\Entity\Attribute\Backend\Datetime',
        'Magento\Catalog\Model\Attribute\Backend\Startdate',
        'Magento\Catalog\Model\Product\Attribute\Backend\Boolean',
        'Magento\Eav\Model\Entity\Attribute\Backend\DefaultBackend',
        'Magento\Catalog\Model\Product\Attribute\Backend\Weight',
        'Magento\Catalog\Model\Product\Attribute\Backend\Price',
    ];`

for attribute to be indexed. I don't get why it matters and why it's hardcoded here. If there are some reasons to do it, please explain it, but if not please fix it. I was trying to find something special in common of these backed models, but no luck - only inheritance after `AbstractBackend`, but don't think it matters. Of course just after deletion of backend model from attribute, search can be done with no problems.

**Comment by afoucret** [comment]
Most backend model contains data that need post processing before being indexed (unserialization most of time). The AttributeData knows how to handle some backed models like ArrayBackend and the other "hardcoded" ones.

To avoid dirty data to be added into the index, we chosed to secure things by limiting indexing of attribute with backend model to those we know how to handle during indexing.

Now it is not completely hardcoded since you can add more backend model in the list through DI :

```xml
<type name="Smile\ElasticsuiteCatalog\Model\Product\Indexer\Fulltext\Datasource\AttributeData">
    <arguments>
        <argument name="indexedBackendModels" xsi:type="array">
            <item name="myCustomBackend" xsi:type="string">\My\Module\Model\Attrribute\Backend\MyBackend.php</item>
        </argument>
    <arguments>
</type>
```

If you chose to do this:
* Value will be read from the one of the catalog_product_entity_* table (int, decimal, text or varchar)
* The text value stored in MySQL will be send to Elasticsearch after it will be processed by the `Smile\ElasticsuiteCatalog\Helper\ProductAttribute::prepareIndexValue` method.
* If you need additional processing like unserialization, you will have to write a plugin for the above method

**TL;DR => it is expected and you have to:**
* explicitly inject your backend model as indexable
* take care of values that will be sent to the engine by yourself

BR,

**Comment by bartek9007** [comment]
Oh, haven't noticed that this array in construct parameter - so, clearly editable. But, you had reasons to do it, hadn't you? Can you tell what is dirty data or give some examples of backend model which should not be present in this array? How can I differ that?

For example we want to make ean number searchable -([https://en.wikipedia.org/wiki/International_Article_Number]((explantation what is ean just in case)). In backend model we validate it against regex and we throw exception if it doesn't match.

**Comment by bartek9007** [comment]
Oh, haven't noticed that this array in construct parameter - so, clearly editable. Can you tell what is dirty data or give some examples of backend model which should not be present in this array? 

For example we want to make ean number searchable (just 13-digits number, but we store it as varchar, to leave possibility to change pattern, just in case). In backend model we validate it against regex and we throw exception if it doesn't match. No additional operations - should this work?
